pkg/adapter/apiserver/delegate.go

```diff
@@ -36,42 +37,48 @@ type resourceDelegate struct {
 var _ cache.Store = (*resourceDelegate)(nil)
 
 func (a *resourceDelegate) Add(obj interface{}) error {
-	event, err := events.MakeAddEvent(a.source, obj, a.ref)
+	ctx, event, err := events.MakeAddEvent(a.source, obj, a.ref)
 	if err != nil {
 		a.logger.Infow("event creation failed", zap.Error(err))
 		return err
 	}
-
-	if result := a.ce.Send(context.Background(), event); !cloudevents.IsACK(result) {
-		a.logger.Errorw("failed to send event", zap.Error(result))
-	}
+	a.sendCloudEvent(ctx, event)
 	return nil
 }
 
 func (a *resourceDelegate) Update(obj interface{}) error {
-	event, err := events.MakeUpdateEvent(a.source, obj, a.ref)
+	ctx, event, err := events.MakeUpdateEvent(a.source, obj, a.ref)
 	if err != nil {
 		a.logger.Info("event creation failed", zap.Error(err))
 		return err
 	}
-
-	if result := a.ce.Send(context.Background(), event); !cloudevents.IsACK(result) {
-		a.logger.Error("failed to send event", zap.Error(result))
-	}
+	a.sendCloudEvent(ctx, event)
 	return nil
 }
 
 func (a *resourceDelegate) Delete(obj interface{}) error {
-	event, err := events.MakeDeleteEvent(a.source, obj, a.ref)
+	ctx, event, err := events.MakeDeleteEvent(a.source, obj, a.ref)
 	if err != nil {
 		a.logger.Info("event creation failed", zap.Error(err))
 		return err
 	}
+	a.sendCloudEvent(ctx, event)
+	return nil
+}
 
-	if result := a.ce.Send(context.Background(), event); !cloudevents.IsACK(result) {
-		a.logger.Error("failed to send event", zap.Error(result))
+// sendCloudEvent sends a cloudevent everytime k8s api event is created, updated or deleted.
+func (a *resourceDelegate) sendCloudEvent(ctx context.Context, event cloudevents.Event) {
+	event = event.Clone()
+	event.SetID(uuid.New().String()) // provide an ID here so we can track it with logging
+	defer a.logger.Debug("Finished sending cloudevent id: ", event.ID())
+	source := event.Context.GetSource()
+	subject := event.Context.GetSubject()
+	a.logger.Debugf("sending cloudevent id: %s, source: %s, subject: %s", event.ID(), source, subject)
```

Comment:
Yes, initially I was trying to write a log for when the cloudevent is sent, but I got persuaded with the logging done in the pingsource and I think the logline "sending cloudevent ..." would do the job, ofcourse if its not sent I'll see the error right after as well. Without these changes, there is no logging for an attempt to send cloudevent at all unless there has been an error. 
or shall I update the code to look like this:
```
if result := a.ce.Send(ctx, event); !cloudevents.IsACK(result) {
		a.logger.Errorw("failed to send cloudevent", zap.Error(result), zap.String("source", source),
			zap.String("subject", subject), zap.String("id", event.ID()))
	} else {
		a.logger.Debugf("cloudevent sent id: %s, source: %s, subject: %s", event.ID(), source, subject)
	}
```
If yes, this could probably go in pingsource as well?

Reply:
with this one more change, the logging will be like "sending cloudevent" -> "cloudevent sent" -> "finished sending cloudevent", which makes sense.

Reply:
Can't we create a shared library or push this into the adapter? 
/cc @lionelvillard

Reply:
I agree this sounds like a good idea. This code is 99% identical to the PingSource code. Do you mind moving it into v2/cloudevents.go (or somewhere else if you feel like it)

Reply:
I don't have the overall idea, but if the code restructuring is the idea, perhaps it should happen with the optimization goal for APIServerSource - the adapter model instead of creating RA per source? Please let me know if I am saying something naive, I am just basing it off with whatever I know.

Reply:
I have worked up on a solution, moved the relevant code to v2/cloudevents.go, let me clean it up quickly, will commit in a bit.